### ISSUE TITLE
bazel doesn't notice when an external repository URL and SHA have changed

### ISSUE THREAD

**Issue by jmhodges** [bazelbuild/bazel#2294]
This is on bazel 0.4.2.

While working on upgrading to Go 1.8beta2, I noticed that when changing my new_http_archive from 

    new_http_archive(
         name = "golang_darwin_amd64",
         build_file_content = "",
         sha256 = "f86c727012e33f3f482b049281aaee24211fe29dfafd121d93f32799a6b6ba2d",
         strip_prefix = "go",
         url = "https://storage.googleapis.com/golang/go1.7.4.darwin-amd64.tar.gz",
     )

to

    new_http_archive(
        name = "golang_darwin_amd64",
        build_file_content = "",
        sha256 = "9cfa3a7064bd201924f53458e1db500957cf2cb7bf87983eb7611e9ced153e35",
        strip_prefix = "go",
        url = "https://storage.googleapis.com/golang/go1.8beta2.darwin-amd64.tar.gz",
    )

that bazel's caching did not "see" the update when attempting to build things that depended on it. Only after using `bazel clean --expunge` did certain expected build failures occur. (There's a `go tool yacc` thing that disappeared in Go 1.8)

This caused some busted commits to make it past my CI that caches bazel output and into master.

**Comment by damienmg** [comment]
This looks like a serious regression.

Which version of Bazel are you using? Can we get a full repro (which build file and a bazel command to run)?

**Comment by kchodorow** [comment]
I can't reproduce this at head (4a099f977df1fd0ca912c31e7acbbe21b0267119).  In a clean repository, I've got:

```
new_http_archive(
    name = "golang",
    build_file_content = "",
    sha256 = "f86c727012e33f3f482b049281aaee24211fe29dfafd121d93f32799a6b6ba2d",
    strip_prefix = "go",
    url = "https://storage.googleapis.com/golang/go1.7.4.darwin-amd64.tar.gz",
)
```

I then build `bazel build @golang//...`, and that generates:

```
$ ls /home/kchodorow/.cache/bazel/_bazel_kchodorow/3789d108650523097f01bbf32a824071/external/golang/
api      blog             CONTRIBUTORS  go1.7.4.darwin-amd64.tar.gz ...
```

Then I change the WS to:

```
new_http_archive(
    name = "golang",
    build_file_content = "",
    sha256 = "9cfa3a7064bd201924f53458e1db500957cf2cb7bf87983eb7611e9ced153e35",
    strip_prefix = "go",
    url = "https://storage.googleapis.com/golang/go1.8beta2.darwin-amd64.tar.gz",
)
```

I run `bazel build @golang//...` again and get:

```
$ ls /home/kchodorow/.cache/bazel/_bazel_kchodorow/3789d108650523097f01bbf32a824071/external/golang/
api      blog             CONTRIBUTORS  go1.8beta2.darwin-amd64.tar.gz ...
```

Can you give some more details about how you're getting this?

**Comment by jmhodges** [comment]
Calling the @golang//... is doing it, but I'm saying it's not happening when you use a target that depends on that repositories. 

I'm getting this with rules_go. I have this is in my WORKSPACE:

    new_http_archive(
        name = "golang_linux_amd64",
        build_file_content = "",
        sha256 = "bb8fe0d81161e4a8b0a8b2145ee5f8a60370baf5d48c07a83f6f09e1ad253bec",
        strip_prefix = "go",
        url = "https://storage.googleapis.com/golang/go1.8rc1.linux-amd64.tar.gz",
    )

    new_http_archive(
        name = "golang_darwin_amd64",
        build_file_content = "",
        sha256 = "faf2530afbc78d8fa64054960a130f10aa2b183abd5879cdfd2ec1a4bca01ff3",
        strip_prefix = "go",
        url = "https://storage.googleapis.com/golang/go1.8rc1.darwin-amd64.tar.gz",
    )

    git_repository(
        name = "io_bazel_rules_go",
        remote = "https://github.com/jmhodges/rules_go.git",
        commit = "1cee2890770a30f4ff3df9eb0cc7ba5ff363c3ae",
    )

    load("@io_bazel_rules_go//go:def.bzl", "go_repositories", "new_go_repository")

    go_repositories(omit_go = True)


The `omit_go` allows me to use the golang repositories I specified instead of rules_go using an older version. And I have this in a BUILD file:

    go_library(
        name = "go_default_library",
        srcs = ["main.go"],
        visibility = ["//visibility:private"],
        deps = [],
    )

    go_binary(
        name = "something_nice",
        library = ":go_default_library",
        visibility = ["//visibility:public"],
    )

And if I change the two golang repositories to another version, and run `bazel build //path/to/something_nice`, the new versions of the Go repositories are not downloaded and used.

**Comment by jmhodges** [comment]
To be clear, I don't think it's the rules_go stuff, and I maybe should have excluded it from my repro. It's really that I don't think calling the repositories exact target explicitly should be required and that calling a target that happens to use that repo should.

**Comment by kchodorow** [comment]
>
> It's really that I don't think calling the repositories exact target
> explicitly should be required and that calling a target that happens to use
> that repo should.
>

Agreed!  Thanks for the detailed info, let me try again to repro.

**Comment by jmhodges** [comment]
Yeah, sorry, I deleted that comment because I wanted to spend more time investigating the rules_go connection, but I'm stumped.

**Comment by kchodorow** [comment]
There are some known issues with Skylark repository invalidation, I'm guessing you're running into something related to that.

**Comment by jmhodges** [comment]
(Ugh, in my repro I included my fork of rules_go but that was just dregs from waiting for some PRs to land. Using the real rules_go fork still demonstrates the problem, though! Editing my comment to correct the io_bazel_rules_go for clarity)

**Comment by kchodorow** [comment]
Repro'd!

**Comment by jmhodges** [comment]
Great! Thank you for your work and patience with me!

**Comment by kchodorow** [comment]
Thanks for the detailed report!

@damienmg: not sure if you want this in for 0.5 or not, it's another instance of Skylark repository rules being cached improperly.  Specifically, in the example above, `omit_go` is changed, but the go rules' `go_repositories` macro isn't re-evaluated.

**Comment by jmhodges** [comment]
(When you say "omit_go is changed", did you mean to write something like "when the golang_*_amd64 repos (that omit_go can be used to skip over defining in rules_go) are changed"? Because I've not changed omit_go since the last `clean --expunge` I did)

**Comment by damienmg** [comment]
Ok thanks for the repro, I'll try to repro it with my recent list of changes. Yes we should do it by 0.5

**Comment by jmhodges** [comment]
Anyone gotten any further with figuring this one out?

**Comment by kchodorow** [comment]
Unfortunately I haven't gotten a chance to yet, but it's still on my radar.

**Comment by jmhodges** [comment]
Thanks for the update!

**Comment by jmhodges** [comment]
I made a git repo just to act as a reproduction. https://github.com/jmhodges/bazel_bug_2294

**Comment by mhlopko** [comment]
Hello Kristina, is this still expected to be fixed in 0.5? It's still at least 2 weeks away, but we're getting there :)

**Comment by damienmg** [comment]
(we are past the deadline actually)

**Comment by damienmg** [comment]
This should probably be a release blocker since it seems like an invalidation issue.

**Comment by jmhodges** [comment]
Annoyingly, the reproduction repo I made stopped building at all because of the `bazelbuild/buildifier` to `bazelbuild/buildtools` rename and when I update the repro to `rules_go` 0.4.3 or 0.3.4, the bug goes away completely.

There is now a `rules_go_0.4.3` and a `rules_go_0.3.4` branch in https://github.com/jmhodges/bazel_bug_2294 that y'all can try out. Happy hunting!

**Comment by damienmg** [comment]
Ok I cannot repro with 0.4.5. Since I fixed several thing in 0.4.5 regarding repository invalidation, I consider this fixed.
